org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/Main.java

```diff
@@ -5377,76 +5286,23 @@ protected void validateOptions(boolean didSpecifyCompliance) {
 			throw new IllegalArgumentException(
 					this.bind("configure.unsupportedWithRelease", version));//$NON-NLS-1$
 		}
-		if (CompilerOptions.VERSION_1_3.equals(version)) {
-			if (!this.didSpecifySource) this.options.put(CompilerOptions.OPTION_Source, CompilerOptions.VERSION_1_3);
-			if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_1);
-		} else if (CompilerOptions.VERSION_1_4.equals(version)) {
-			if (this.didSpecifySource) {
-				Object source = this.options.get(CompilerOptions.OPTION_Source);
-				if (CompilerOptions.VERSION_1_3.equals(source)) {
-					if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_2);
-				} else if (CompilerOptions.VERSION_1_4.equals(source)) {
-					if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_4);
-				}
-			} else {
-				this.options.put(CompilerOptions.OPTION_Source, CompilerOptions.VERSION_1_3);
-				if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_2);
-			}
-		} else if (CompilerOptions.VERSION_1_5.equals(version)) {
+		if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version)) {
 			if (this.didSpecifySource) {
-				Object source = this.options.get(CompilerOptions.OPTION_Source);
-				if (CompilerOptions.VERSION_1_3.equals(source)
-						|| CompilerOptions.VERSION_1_4.equals(source)) {
-					if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_4);
-				} else if (CompilerOptions.VERSION_1_5.equals(source)) {
-					if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_5);
-				}
-			} else {
-				this.options.put(CompilerOptions.OPTION_Source, CompilerOptions.VERSION_1_5);
-				if (!this.didSpecifyTarget) this.options.put(CompilerOptions.OPTION_TargetPlatform, CompilerOptions.VERSION_1_5);
+				throw new IllegalArgumentException(this.bind("configure.unsupportedSourceVersion", //$NON-NLS-1$
+						this.options.get(CompilerOptions.OPTION_Source), CompilerOptions.getFirstSupportedJavaVersion()));
 			}
```

Comment:
What about to leave only the compliance check here? I believe I've added that looking around at the code, but now I believe it should be sufficient to only check compliance argument here and leave other checks be done below.

Reply:
Actually upon further debugging, I feel that `if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version)) {` can never be true.

@iloveeclipse can you please check this ? 

Also: 

Is `configure.unsupportedSourceVersion` duplicating `configure.source` ? 

Is `configure.unsupportedTargetVersion` duplicating `configure.targetJDK ` ??

Reply:
> I feel that if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version)) { can never be true because we would have already rejected the option string at the caller itself ?

I have same feeling. I've added `if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version))` and below checks to be "compatible" with the surrounding checks.

TBH, ~1.300 lines of spaghetti code in the `Main.configure(String[])` method is nothing I enjoy to study

Reply:
> Is configure.unsupportedSourceVersion duplicating configure.source ?
> Is configure.unsupportedTargetVersion duplicating configure.targetJDK  ??

What do you mean by duplicating? I do not see duplicates of any kind here?

Reply:
Running the batch compiler with these options: `--release 5 X.java` results in:

```
Failed to init Classpath for JEP 247 for JDK C:\Program Files\Java\jdk-21\jdk-21\lib\jrt-fs.jar
java.io.IOException: No provider for uri jar:file:/C:/Program%20Files/Java/jdk-21/jdk-21/lib/ct.sym
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathJep247Jdk12.initialize(ClasspathJep247Jdk12.java:125)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\OneDrive - Advantest\Desktop\Eclipse SDKs\I20240626-1800\eclipse\plugins\org.eclipse.pde.api.tools.annotations_1.3.0.v20240207-2106.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/OneDrive%20-%20Advantest/Desktop/Eclipse%20SDKs/I20240626-1800/eclipse/plugins/org.eclipse.pde.api.tools.annotations_1.3.0.v20240207-2106.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\OneDrive - Advantest\Desktop\Eclipse SDKs\I20240626-1800\eclipse\plugins\org.osgi.annotation.versioning_1.1.2.202109301733.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/OneDrive%20-%20Advantest/Desktop/Eclipse%20SDKs/I20240626-1800/eclipse/plugins/org.osgi.annotation.versioning_1.1.2.202109301733.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\OneDrive - Advantest\Desktop\Eclipse SDKs\I20240626-1800\eclipse\plugins\org.osgi.annotation.bundle_2.0.0.202202082230.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/OneDrive%20-%20Advantest/Desktop/Eclipse%20SDKs/I20240626-1800/eclipse/plugins/org.osgi.annotation.bundle_2.0.0.202202082230.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\OneDrive - Advantest\Desktop\Eclipse SDKs\I20240626-1800\eclipse\plugins\org.osgi.service.component.annotations_1.5.1.202212101352.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/OneDrive%20-%20Advantest/Desktop/Eclipse%20SDKs/I20240626-1800/eclipse/plugins/org.osgi.service.component.annotations_1.5.1.202212101352.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\OneDrive - Advantest\Desktop\Eclipse SDKs\I20240626-1800\eclipse\plugins\org.osgi.service.metatype.annotations_1.4.1.202109301733.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/OneDrive%20-%20Advantest/Desktop/Eclipse%20SDKs/I20240626-1800/eclipse/plugins/org.osgi.service.metatype.annotations_1.4.1.202109301733.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\git\eclipse.jdt.core\org.eclipse.jdt.core.compiler.batch\lib\javax20api.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/git/eclipse.jdt.core/org.eclipse.jdt.core.compiler.batch/lib/javax20api.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
Failed to init Classpath for jar file C:\Users\srikanth.sankaran\OneDrive - Advantest\Desktop\Eclipse SDKs\I20240626-1800\eclipse\plugins\org.apache.ant_1.10.14.v20230922-1200\lib\ant.jar
java.io.IOException: No provider for uri jar:file:/C:/Users/srikanth.sankaran/OneDrive%20-%20Advantest/Desktop/Eclipse%20SDKs/I20240626-1800/eclipse/plugins/org.apache.ant_1.10.14.v20230922-1200/lib/ant.jar
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:191)
	at org.eclipse.jdt.internal.compiler.batch.ClasspathMultiReleaseJar.initialize(ClasspathMultiReleaseJar.java:40)
	at org.eclipse.jdt.internal.compiler.batch.FileSystem.<init>(FileSystem.java:235)
	at org.eclipse.jdt.internal.compiler.batch.Main.getLibraryAccess(Main.java:3400)
	at org.eclipse.jdt.internal.compiler.batch.Main.performCompilation(Main.java:4678)
	at org.eclipse.jdt.internal.compiler.batch.Main.compile(Main.java:1711)
	at org.eclipse.jdt.internal.compiler.batch.Main.main(Main.java:1477)
Caused by: java.nio.file.ProviderNotFoundException: Provider "jar" not found
	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:238)
	at org.eclipse.jdt.internal.compiler.util.JRTUtil.getJarFileSystem(JRTUtil.java:182)
	... 6 more
----------
1. ERROR in C:\tmp\X.java (at line 0)
	class Test {
	^
The Java version '1.5' is no longer supported. Use 1.8 or later.
----------
1 problem (1 error)
```
I think the case arm `case INSIDE_RELEASE:` should be enhanced with some validations as is done for `case INSIDE_SOURCE :` and `case INSIDE_TARGET`

Reply:
Are you sure that "--release 5" was supported before? If I remember right, 1.7 was minimal version for release.

Reply:
> What about to leave only the compliance check here? I believe I've added that looking around at the code, but now I believe it should be sufficient to only check compliance argument here and leave other checks be done below.

I believe all these blocks are dead:

`if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version)) {` which features under `if (didSpecifyCompliance) {`
`if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version)) {` which features under `else if (this.didSpecifySource) {` and 
`if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(targetVersion)) {` which features under `if (this.didSpecifyTarget) {`

Because the didSpecify flags are set only when -source -target et al are passed. We already throw if we don't recognize the versions. See `case INSIDE_SOURCE :` etc.

If we get rid of these dead blocks of code, we will realize that `configure.unsupportedTargetVersion` has been merely duplicating `configure.targetJDK` and likewise configure.source and its doppleganger.

This is not a problrm with the PR, but predates

Reply:
> > Is configure.unsupportedSourceVersion duplicating configure.source ?
> > Is configure.unsupportedTargetVersion duplicating configure.targetJDK  ??
> 
> What do you mean by duplicating? I do not see duplicates of any kind here?

See my comment elsewhere about dead blocks in `org.eclipse.jdt.internal.compiler.batch.Main.validateOptions(boolean)` if those are pruned, we will realize these messages are issued under identical circumstances

Reply:
> > I feel that if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version)) { can never be true because we would have already rejected the option string at the caller itself ?
> 
> I have same feeling. I've added `if(CompilerOptions.UNSUPPORTED_VERSIONS.contains(version))` and below checks to be "compatible" with the surrounding checks.
> 
> TBH, ~1.300 lines of spaghetti code in the `Main.configure(String[])` method is nothing I enjoy to study

I hear you! I have carefully verified that `org.eclipse.jdt.internal.compiler.batch.Main.validateOptions(boolean)` and in general `org.eclipse.jdt.internal.compiler.batch.Main.java` changes proposed preserve the existing semantics - even where they are suspect/dubious. So that is good!

Reply:
Regarding errors reported: I can't reproduce.
```
/usr/lib/jvm/java-21/bin/java -jar org.eclipse.jdt.core.compiler.batch-3.39.0-SNAPSHOT.jar --release 5 /data/4x_platform_workspace/A/src/X.java

release 5 is not found in the system
```